File: testing16-selenium-pom/src/main/java/pages/LoginPage.java
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public class LoginPage extends BasePage {
//    <input
//        data-v-1f99f73c=""
//        class="oxd-input oxd-input--active"
//        name="username"
//        placeholder="Username"
//        autofocus="">
//    1. tìm tên thẻ: input, button, a, p, h1, h2,...
//    2. tìm theo class name
//    3. tìm theo id, placeholder, name,...
//    4. tìm theo quan hệ cha-con
//    ưu tiên cao nhất: tìm theo id -> name -> placeholder -> class name ->
    private static final By USERNAME_INPUT = By.xpath("//input[@name='username']");
    private static final By PASSWORD_INPUT = By.xpath("//input[@name='password']");
    private static final By LOGIN_BUTTON = By.xpath("//button[@type='submit']");

    public LoginPage(WebDriver driver, WebDriverWait wait) {
        super(driver, wait);
    }

//    define các step
    public void open() {
        String url = "https://opensource-demo.orangehrmlive.com/web/index.php/auth/login";
        driver.get(url);
        wait.until(ExpectedConditions.visibilityOfElementLocated(USERNAME_INPUT));
    }

    public void enterUsername(String username) throws InterruptedException {
        WebElement usernameInput = driver.findElement(USERNAME_INPUT);
        highlight(usernameInput);

        usernameInput.sendKeys(username);
        Thread.sleep(2000);
        unhighlight(usernameInput);
    }

    public void enterPassword(String password)throws InterruptedException {
        WebElement passwordInput = driver.findElement(PASSWORD_INPUT);
        highlight(passwordInput);

        passwordInput.sendKeys(password);
        Thread.sleep(2000);
        unhighlight(passwordInput);
    }

    public void clickLoginBtn() throws InterruptedException {
        WebElement loginBtn = driver.findElement(LOGIN_BUTTON);
        highlight(loginBtn);

        loginBtn.click();

        Thread.sleep(1000);
        unhighlight(loginBtn);
    }

//    func gom 3 step này với nhau
    public void login(String username, String password) throws  InterruptedException {
        open();
        enterUsername(username);
        enterPassword(password);
        clickLoginBtn();
    }
}

File: testing16-playwright/pages/LoginPage.ts
import { Locator, Page } from "@playwright/test";
import { highlight, unhighlight } from "./utils/highlight";

export class LoginPage {
    // thuộc tính để tương tác với browser
    readonly page: Page
    readonly usernameInput: Locator
    readonly passwordInput: Locator
    readonly loginBtn: Locator

    // <input
    //     name="username"
    //     type="text"
    //     placeholder="Username"
    //     class="oxd-input oxd-input--focus"
    // >

    constructor(page: Page) {
        this.page = page
        this.usernameInput = page.locator("input[name='username']")
        // C2: xpath
        // this.usernameInput = page.locator("//input[@name='username']")
        // C3: getByRole
        // this.usernameInput = page.getByRole('textbox', { name: 'Username' })
        // C4: getByPlaceholder
        // this.usernameInput = page.getByPlaceholder('Username')
        // C5: getyText, getByLabel,...
        this.passwordInput = page.locator("//input[@name='password']")
        this.loginBtn = page.getByRole('button', { name: 'Login' })
    }

    // define các step thì dùng BẤT ĐỒNG BỘ (async - await)
    async login(username: string, password: string): Promise<void> {
        // B1: truy cập vào trang login
        await this.page.goto("https://opensource-demo.orangehrmlive.com/web/index.php/auth/login", {
            // domcontentloaded: trang load xong DOM, nhưng chưa load xong các resource (css, js, image,...)
            waitUntil: "domcontentloaded",
            timeout: 30000
        })

        // B2: nhập username
        highlight(this.usernameInput)
        await this.usernameInput.fill(username)
        await this.page.waitForTimeout(1000)
        unhighlight(this.usernameInput)
        
        // B3: nhập password
        highlight(this.passwordInput)
        await this.passwordInput.fill(password)
        await this.page.waitForTimeout(1000)
        unhighlight(this.passwordInput)

        // B4: click login button
        highlight(this.loginBtn)
        await this.loginBtn.click()
        await this.page.waitForTimeout(1000)
        unhighlight(this.loginBtn)
    }

    async isLoginSuccessful(): Promise<boolean> {
        try {
            // endpoint không chứa từ "login" => login thành công
            // bên trong waitForURL() có thể truyền vào 1 string, regex, hoặc function
            // refer regular expression
            // .* => match tất cả các ký tự
            await this.page.waitForURL(/.*dashboard/, {
                waitUntil: "domcontentloaded",
                timeout: 30000
            })
            return true
        } catch (error) {
            return false
        }
    }
}